In [1]:
import pandas as pd
import numpy as np

In [2]:
# Load all datasets

circuits = pd.read_csv("../data/raw/circuits.csv")
constructors = pd.read_csv("../data/raw/constructors.csv")
constructor_results = pd.read_csv("../data/raw/constructor_results.csv")
constructor_standings = pd.read_csv("../data/raw/constructor_standings.csv")
drivers = pd.read_csv("../data/raw/drivers.csv")
driver_standings = pd.read_csv("../data/raw/driver_standings.csv")
lap_times = pd.read_csv("../data/raw/lap_times.csv")
pit_stops = pd.read_csv("../data/raw/pit_stops.csv")
qualifying = pd.read_csv("../data/raw/qualifying.csv")
races = pd.read_csv("../data/raw/races.csv")
results = pd.read_csv("../data/raw/results.csv")
seasons = pd.read_csv("../data/raw/seasons.csv")
sprint_results = pd.read_csv("../data/raw/sprint_results.csv")
status = pd.read_csv("../data/raw/status.csv")

In [3]:
datasets = {
    "circuits": circuits,
    "constructors": constructors,
    "constructor_results": constructor_results,
    "constructor_standings": constructor_standings,
    "drivers": drivers,
    "driver_standings": driver_standings,
    "lap_times": lap_times,
    "pit_stops": pit_stops,
    "qualifying": qualifying,
    "races": races,
    "results": results,
    "seasons": seasons,
    "sprint_results": sprint_results,
    "status": status
}

In [4]:
for name, df in datasets.items():
    print(f"{name}: {df.shape}")

circuits: (78, 9)
constructors: (214, 5)
constructor_results: (12986, 5)
constructor_standings: (13752, 7)
drivers: (865, 9)
driver_standings: (35603, 7)
lap_times: (878504, 6)
pit_stops: (22548, 7)
qualifying: (11212, 9)
races: (1171, 18)
results: (27480, 18)
seasons: (77, 2)
sprint_results: (568, 17)
status: (141, 2)


In [5]:
circuits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78 entries, 0 to 77
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   circuitId   78 non-null     int64  
 1   circuitRef  78 non-null     object 
 2   name        78 non-null     object 
 3   location    78 non-null     object 
 4   country     78 non-null     object 
 5   lat         78 non-null     float64
 6   lng         78 non-null     float64
 7   alt         78 non-null     int64  
 8   url         78 non-null     object 
dtypes: float64(2), int64(2), object(5)
memory usage: 5.6+ KB


In [6]:
drivers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 865 entries, 0 to 864
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   driverId     865 non-null    int64 
 1   driverRef    865 non-null    object
 2   number       865 non-null    object
 3   code         865 non-null    object
 4   forename     865 non-null    object
 5   surname      865 non-null    object
 6   dob          865 non-null    object
 7   nationality  865 non-null    object
 8   url          865 non-null    object
dtypes: int64(1), object(8)
memory usage: 60.9+ KB


In [7]:
drivers.isnull().sum()

driverId       0
driverRef      0
number         0
code           0
forename       0
surname        0
dob            0
nationality    0
url            0
dtype: int64

In [8]:
for name, df in datasets.items():
    print("=" * 50)
    print(f"Dataset: {name}")
    print(df.isnull().sum())
    print()

Dataset: circuits
circuitId     0
circuitRef    0
name          0
location      0
country       0
lat           0
lng           0
alt           0
url           0
dtype: int64

Dataset: constructors
constructorId     0
constructorRef    0
name              0
nationality       0
url               0
dtype: int64

Dataset: constructor_results
constructorResultsId    0
raceId                  0
constructorId           0
points                  0
status                  0
dtype: int64

Dataset: constructor_standings
constructorStandingsId    0
raceId                    0
constructorId             0
points                    0
position                  0
positionText              0
wins                      0
dtype: int64

Dataset: drivers
driverId       0
driverRef      0
number         0
code           0
forename       0
surname        0
dob            0
nationality    0
url            0
dtype: int64

Dataset: driver_standings
driverStandingsId    0
raceId               0
driverId          

### Missing Values Check

- Checked all 14 datasets using `isnull().sum()`.
- No missing values were found in any dataset.
- No imputation or row deletion is required.

In [9]:
for name, df in datasets.items():
    print(f"{name}: {df.duplicated().sum()} duplicate rows")

circuits: 0 duplicate rows
constructors: 0 duplicate rows
constructor_results: 0 duplicate rows
constructor_standings: 0 duplicate rows
drivers: 0 duplicate rows
driver_standings: 0 duplicate rows
lap_times: 1707 duplicate rows
pit_stops: 0 duplicate rows
qualifying: 0 duplicate rows
races: 0 duplicate rows
results: 0 duplicate rows
seasons: 0 duplicate rows
sprint_results: 0 duplicate rows
status: 0 duplicate rows


### Duplicate Check

- Checked duplicates in all datasets.
- Only `lap_times` contained duplicate rows.
- Found **1,707** complete duplicate records.
- Verified the duplicates before removing them.

In [10]:
lap_times[lap_times.duplicated()]

,raceId,driverId,lap,position,time,milliseconds
213318,372,77,1,2,1:42.315,102315
213319,372,77,2,2,1:34.782,94782
213321,372,77,4,2,1:35.886,95886
213322,372,77,5,2,1:36.239,96239
213323,372,77,6,2,1:36.252,96252
...,...,...,...,...,...,...
233348,356,158,33,9,1:38.255,98255
233349,356,158,34,9,1:38.258,98258
233350,356,158,35,9,1:38.502,98502
233351,356,158,36,9,1:41.150,101150


In [11]:
lap_times.shape

(878504, 6)

In [12]:
lap_times.drop_duplicates().shape

(876797, 6)

In [13]:
lap_times[lap_times.duplicated(keep=False)]

,raceId,driverId,lap,position,time,milliseconds
100393,372,137,1,4,1:44.692,104692
100394,372,137,2,4,1:37.066,97066
100395,372,137,3,4,1:36.518,96518
100396,372,137,4,4,1:36.807,96807
100397,372,137,5,4,1:36.700,96700
...,...,...,...,...,...,...
233348,356,158,33,9,1:38.255,98255
233349,356,158,34,9,1:38.258,98258
233350,356,158,35,9,1:38.502,98502
233351,356,158,36,9,1:41.150,101150


In [14]:
lap_times[lap_times.duplicated(keep=False)]

,raceId,driverId,lap,position,time,milliseconds
100393,372,137,1,4,1:44.692,104692
100394,372,137,2,4,1:37.066,97066
100395,372,137,3,4,1:36.518,96518
100396,372,137,4,4,1:36.807,96807
100397,372,137,5,4,1:36.700,96700
...,...,...,...,...,...,...
233348,356,158,33,9,1:38.255,98255
233349,356,158,34,9,1:38.258,98258
233350,356,158,35,9,1:38.502,98502
233351,356,158,36,9,1:41.150,101150


In [15]:
lap_times = lap_times.drop_duplicates()

In [16]:
lap_times.shape

(876797, 6)

### Duplicate Check

- Checked duplicates in all datasets.
- Only `lap_times` contained duplicate records.
- Found 1,707 complete duplicate rows.
- Verified the duplicates before removing them.
- Removed the duplicate rows successfully.

In [17]:
for name, df in datasets.items():
    print("=" * 60)
    print(f"Dataset: {name}")
    print(df.dtypes)
    print()

Dataset: circuits
circuitId       int64
circuitRef     object
name           object
location       object
country        object
lat           float64
lng           float64
alt             int64
url            object
dtype: object

Dataset: constructors
constructorId      int64
constructorRef    object
name              object
nationality       object
url               object
dtype: object

Dataset: constructor_results
constructorResultsId      int64
raceId                    int64
constructorId             int64
points                  float64
status                   object
dtype: object

Dataset: constructor_standings
constructorStandingsId      int64
raceId                      int64
constructorId               int64
points                    float64
position                   object
positionText               object
wins                        int64
dtype: object

Dataset: drivers
driverId        int64
driverRef      object
number         object
code           object
forename      

In [19]:
pd.to_datetime(races["date"])

0      2009-03-29
1      2009-04-05
2      2009-04-19
3      2009-04-26
4      2009-05-10
          ...    
1166   2026-11-01
1167   2026-11-08
1168   2026-11-22
1169   2026-11-29
1170   2026-12-06
Name: date, Length: 1171, dtype: datetime64[ns]

In [21]:
races["fp1_date"].unique()

array(['\\N', '2021-04-16', '2022-03-18', '2021-03-26', '2021-11-19',
       '2021-04-30', '2021-05-07', '2021-05-21', '2021-06-04',
       '2021-06-25', '2021-06-18', '2021-07-02', '2021-07-16',
       '2021-07-30', '2021-08-27', '2021-09-03', '2021-09-10',
       '2021-09-24', '2021-10-08', '2021-10-22', '2021-11-05',
       '2021-11-12', '2021-12-03', '2021-12-10', '2022-03-25',
       '2022-04-08', '2022-04-22', '2022-05-06', '2022-05-20',
       '2022-05-27', '2022-06-10', '2022-06-17', '2022-07-01',
       '2022-07-08', '2022-07-22', '2022-07-29', '2022-08-26',
       '2022-09-02', '2022-09-09', '2022-09-30', '2022-10-07',
       '2022-10-21', '2022-10-28', '2022-11-11', '2022-11-18',
       '2023-03-03', '2023-03-17', '2023-03-31', '2023-04-28',
       '2023-05-05', '2023-05-26', '2023-06-02', '2023-06-16',
       '2023-06-30', '2023-07-07', '2023-07-21', '2023-07-28',
       '2023-08-25', '2023-09-01', '2023-09-15', '2023-09-22',
       '2023-10-06', '2023-10-20', '2023-10-27',

In [22]:
races["fp1_date"].value_counts(dropna=False).head(10)

fp1_date
\N            1035
2025-04-11       1
2024-11-21       1
2024-11-29       1
2024-12-06       1
2025-03-14       1
2025-03-21       1
2025-04-04       1
2025-04-18       1
2024-03-22       1
Name: count, dtype: int64

In [23]:
# Replace \N with actual missing values
date_columns = [
    "date",
    "fp1_date",
    "fp2_date",
    "fp3_date",
    "quali_date",
    "sprint_date"
]

for col in date_columns:
    races[col] = races[col].replace("\\N", pd.NA)
    races[col] = pd.to_datetime(races[col])

In [24]:
races.dtypes

raceId                  int64
year                    int64
round                   int64
circuitId               int64
name                   object
date           datetime64[ns]
time                   object
url                    object
fp1_date       datetime64[ns]
fp1_time               object
fp2_date       datetime64[ns]
fp2_time               object
fp3_date       datetime64[ns]
fp3_time               object
quali_date     datetime64[ns]
quali_time             object
sprint_date    datetime64[ns]
sprint_time            object
dtype: object

In [25]:
drivers["dob"] = pd.to_datetime(drivers["dob"])

In [26]:
drivers.dtypes

driverId                int64
driverRef              object
number                 object
code                   object
forename               object
surname                object
dob            datetime64[ns]
nationality            object
url                    object
dtype: object

# Data Cleaning Summary

The following data quality checks and transformations were performed:

### ✅ Missing Values
- Checked all 14 datasets.
- No missing values were found.

### ✅ Duplicate Records
- Checked all datasets for duplicate rows.
- Found 1,707 duplicate rows in the `lap_times` dataset.
- Verified that they were complete duplicates.
- Removed the duplicate rows.

### ✅ Data Type Validation
- Verified data types for all datasets.
- Converted the following columns to `datetime`:
  - `drivers.dob`
  - `races.date`
  - `races.fp1_date`
  - `races.fp2_date`
  - `races.fp3_date`
  - `races.quali_date`
  - `races.sprint_date`

### ✅ Special Value Handling
- Replaced `\N` values in race date columns with missing values (`NA`) before datetime conversion.

### ✅ Final Status
The datasets are cleaned, validated, and ready to be exported for SQL analysis.

In [27]:
# Update the cleaned DataFrames in the dictionary
datasets["drivers"] = drivers
datasets["races"] = races
datasets["lap_times"] = lap_times

# Export all cleaned datasets
for name, df in datasets.items():
    df.to_csv(f"../data/cleaned/{name}.csv", index=False)

print("🎉 All cleaned datasets exported successfully!")

🎉 All cleaned datasets exported successfully!
